core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -681,6 +694,17 @@ class GroupCoordinator(val brokerId: Int,
   def onCompleteJoin(group: GroupMetadata) {
     var delayedStore: Option[DelayedStore] = None
     group synchronized {
+      if (group.generationId == 0) {
+        if (group.newMemberAdded && time.milliseconds() < group.initialRebalanceTimeout) {
+          group.newMemberAdded = false
+          val delay = scala.math.min(groupConfig.groupInitialRebalanceDelayMs, group.initialRebalanceTimeout - time.milliseconds()).toInt
+          val groupKey = GroupKey(group.groupId)
+          joinPurgatory.watchOperation(new DelayedJoin(this, group, delay, group.initialRebalanceTimeout), Seq(groupKey))
```

Comment:
> On the empty group we add a FurtherDelayedJoin to the purgatory that is scheduled to run in delayMs. In the meantime another member gets added to the group.

I think `prepareRebalance` will only be called once, because from either `addMemberAndRebalance` or `updateMemberAndRebalance` we call `maybePrepareRebalance`, and if the group is already in `prepareRebalance` state we won't call `prepareRebalance` again, so only on delayed operation will be added still.

> one issue with this is that the total delay is going to be min(numMembers * configured-value, rebalanceTimeout)

Hmm, this seems a bit different from what I understood from KIP-134 description. Here's my understanding: say the rebalance timeout is 10 sec, and delay is 3 sec. Then after the first joined member of an empty group:

1. If there are no more member joining within 3 sec, we will complete `prepareRebalance` at time 3.
2. If there are `one or more` members joined within 3 sec, we will extend to recheck at time 6; and if there are no more member joining from time 3 - 6, we complete at time 6.
3. And so on. But we will still at latest complete at time 10. I.e. if there are no more member joining from time 9 - 10, we complete at time 10.